Zagrios/bs-manager
# feat: respect system proxy settings on Windows platform

## Scope

[closes TICKET-702](https://github.com/Zagrios/bs-manager/issues/702)

## Implementation

Makes BSManager respect system proxy settings on Windows.

1. Detect if BSManager is running on Windows. If it does, execute following steps.
2. Read Windows system proxy settings from registry by using `Winreg` package.
3. Force a global proxy by using `global-agent` package.

I am really not familiar to NodeJS, so implementation code should be ugly. Feel free to edit the code or give any suggestions.

I do not have Linux and macOS and is unlikely to use them. This implementation does not cover Linux or macOS.

## How to Test

Without system proxy enabled on Windows:

- Run BSManager, all functions works fine like before.
- In the log of BSManager, you should see `Proxy Enabled: false`.

With system proxy enabled on Windows:

- Set system proxy settings and enable it on Windows.
- Run BSManager.
- In the log of BSManager, you should see:
  ```log
  [2024-12-18 17:08:04.386] [info]  Proxy Server:  http://127.0.0.1:7890
  [2024-12-18 17:08:04.403] [info]  Proxy Override:  localhost;127.*;192.168.*;10.*;172.16.*;172.17.*;172.18.*;172.19.*;172.20.*;172.21.*;172.22.*;172.23.*;172.24.*;172.25.*;172.26.*;172.27.*;172.28.*;172.29.*;172.30.*;172.31.*;<local>
  [2024-12-18 17:08:04.403] [info]  Using system proxy: http://127.0.0.1:7890
  ```
  That means BSManager is using system proxy.
- If your have access to the proxy and can see which traffic goes through it, you can see BSManager's traffic goes through it.



## Emoji Guide

**For reviewers: Emojis can be added to comments to call out blocking versus non-blocking feedback.**

E.g: Praise, minor suggestions, or clarifying questions that don’t block merging the PR.

> 🟢 Nice refactor!

> 🟡 Why was the default value removed?

E.g: Blocking feedback must be addressed before merging.

> 🔴 This change will break something important

|              |                |                                     |
| ------------ | -------------- | ----------------------------------- |
| Blocking     | 🔴 ❌ 🚨       | RED                                 |
| Non-blocking | 🟡 💡 🤔 💭    | Yellow, thinking, etc               |
| Praise       | 🟢 💚 😍 👍 🙌 | Green, hearts, positive emojis, etc |
